### PR TITLE
[otbn] add ECDSA signature generation subroutine

### PR DESCRIPTION
**Pull request by felixmiller**
Adds subroutine for ECDSA signature generation + test

This finishes #5940

### CODE REVIEW

File: sw/otbn/code-snippets/p384.s [hunk 1]
```diff
@@ -1372,52 +1372,300 @@ scalar_mult_int_p384:
...
+  bn.xor    w31, w31, w31
 
-  jal      x1, scalar_mult_int_p384
+  jal       x1, scalar_mult_int_p384
+
+  ret
+
+
+/**
+ * Variable time modular multiplicative inverse computation
```
Comment (imphil) [inline sw/otbn/code-snippets/p384.s:1426]:
```suggestion
 * Variable-time modular multiplicative inverse computation
```

File: sw/otbn/code-snippets/p384.s [hunk 2]
```diff
@@ -1372,52 +1372,300 @@ scalar_mult_int_p384:
...
+ *
+ * returns x_inv = x^-1 mod m
+ *
+ * This routine computes the modular multiplicative inverse for any x < m in
+ * the finite field GF(m) where m is prime.
+ *
+ * For inverse computation, Fermat's little theorem is used, i.e.
+ * we compute x^-1 = x^(m-2) mod m.
+ * For exponentiation we use a standard, variable time (!) square and multiply
```
Comment (imphil) [inline sw/otbn/code-snippets/p384.s:1435]:
```suggestion
 * For exponentiation we use a standard, variable-time (!) square and multiply
```

File: sw/otbn/code-snippets/p384.s [hunk 3]
```diff
@@ -1372,52 +1372,300 @@ scalar_mult_int_p384:
...
+ *
+ * This routine is mainly intended to be used for inversion of scalars in
+ * context of the P-384 curve. In theory, it can be used with any 384-bit
+ * modulus m with a corresponding 385-bit Barrett constant u,
+ * where u[383:192] = 0.
+ *
+ * Note: When used for P-384 scalar inversion, the routine will need 672 calls
+ * to the multiplication routine. By using an adder chain this could be reduced
+ * to ~433 multiplications, however, for the cost of a significant increase of
+ * codes size.
```
Comment (imphil) [inline sw/otbn/code-snippets/p384.s:1446]:
```suggestion
 * to ~433 multiplications, however, at the cost of a significant code size increase.
```

File: sw/otbn/code-snippets/p384.s [hunk 4]
```diff
@@ -1372,52 +1372,300 @@ scalar_mult_int_p384:
...
+ *
+ * @param[in]  [w13, w12]: m, 384 bit modulus
+ * @param[in]  [w15, w14]: u[383:0], lower 384 bit of pre-computed Barrett
+ *                          constant corresponding to modulus m
+ * @param[in]  [w30, w29]: x, 384 bit operand
+ * @param[in]  w31, all-zero
+ * @param[out] [w17, w16]: x_inv, modular multiplicative inverse
+ *
+ * Flags: Flags when leaving this subroutine depend on a potentially discarded
+ *        value and therefore are not usable after return.
```
Comment (imphil) [inline sw/otbn/code-snippets/p384.s:1459]: Could you rephrase that sentence, I'm not sure I understand it. How can a flag depend on another value? Do you mean the flag has no meaning beyond the scope of this subroutine?

File: sw/otbn/code-snippets/p384.s [hunk 5]
```diff
@@ -1372,52 +1372,300 @@ scalar_mult_int_p384:
...
+    andi      x2, x2, 1
+    beq       x2, x0, nomul
+
+    /* multiply: [w17,w16] <= [w17, w16]*[w30,w29] mod [w13, w12] */
+    bn.mov    w10, w29
+    bn.mov    w11, w30
+    jal       x1, barrett384_p384
+
+    nomul:
+    nop
```
Comment (imphil) [inline sw/otbn/code-snippets/p384.s:1501]: indent? (not sure if that makes sense or is consistent)

File: sw/otbn/code-snippets/p384.s [hunk 6]
```diff
@@ -1372,52 +1372,300 @@ scalar_mult_int_p384:
...
+  la        x20, p384_gx
+
+  /* set dmem pointer to basepoint y-coordinate */
+  la        x21, p384_gy
+
+  /* set dmem pointer to secret random scalar k */
+  la        x19, dptr_k
+  lw        x19, 0(x19)
+
+    /* set pointer to blinding parameter */
```
Comment (imphil) [inline sw/otbn/code-snippets/p384.s:1556]: indent

**Review by imphil — APPROVED** [review]
LGTM, thanks!

File: sw/otbn/code-snippets/p384.s (hunk 4 above)
Comment (felixmiller) [reply sw/otbn/code-snippets/p384.s:1459]: It is often the case that flags depend on the result of last computation in the subroutine (and by this on the result), so one could think of using the M/L/Z flags in a follow-up computation. But often this final computation is a conditional subtraction that is executed in any case (for the sake of constant time) but then potentially discarded. I agree, your suggestion (no meaning beyond the scope...) is clearer. Changed this here and at other places as well.

File: sw/otbn/code-snippets/p384.s (hunk 5 above)
Comment (felixmiller) [reply sw/otbn/code-snippets/p384.s:1501]: Actually, so far, I didn't indent after label, just b/c a block of follow-up instruction is not always well-defined (not always clear where it ends). So for the sake of consistency I would rather not indent here. Let me know if you see differently.

Comment (imphil) [reply sw/otbn/code-snippets/p384.s:1496]: Keeping it as-is is fine!
